### PR TITLE
Fix/peer reconnection

### PR DESCRIPTION
**Pull request by munna0908**
# Description

This PR fixes the issue that the non bootnodes keep disconnected after loosing all connections. 
As part of this change, every node initiates a new routine which periodically(Every 30 sec)attempts to make new connection, if peercount is less than the minimumPeercount (currently set to 1 )

# Changes include

- [x] Bugfix (non-breaking change that solves an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have tested this code
- [ ] I have updated the README and other relevant documents (guides...)
- [ ] I have added sufficient documentation both in code, as well as in the READMEs

### CODE REVIEW

File: network/server_test.go [hunk 1]
```diff
@@ -212,6 +214,58 @@ func TestNat(t *testing.T) {
 	})
 }
 
+func WaitUntilPeerReconnects(ctx context.Context, srv *Server, id peer.ID) (bool, error) {
```
Comment (Kourin1996) [inline network/server_test.go]: I think `WaitUntilPeerConnectsTo` is better? because you don't check node disconnects from some peer in this function.

File: protocol/syncer.go [hunk 2]
```diff
@@ -484,9 +490,11 @@ func (s *Syncer) WatchSyncWithPeer(p *syncPeer, handler func(b *types.Block) boo
...
 
 		b := p.popBlock()
-		if err := s.blockchain.WriteBlocks([]*types.Block{b}); err != nil {
-			s.logger.Error("failed to write block", "err", err)
-			break
+		if b != nil {
+			if err := s.blockchain.WriteBlocks([]*types.Block{b}); err != nil {
+				s.logger.Error("failed to write block", "err", err)
+				break
+			}
```
Comment (Kourin1996) [inline protocol/syncer.go]: `handler` may expect a block, so I think you shouldn't pass nil to handler. So you might want to break if popBlock returns nil

File: network/server.go [hunk 3]
```diff
@@ -182,6 +185,27 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
 	return srv, nil
 }
 
+// checkPeerCount will attempt to make new connections if the active peer count is lesser than the specified limit.
+func (s *Server) checkPeerConnections() {
+	for {
+
+		select {
+		case <-time.After(30 * time.Second):
```
Comment (Kourin1996) [inline network/server.go:193]: I'm sure polling approach will work. So it's just curious question => is it possible for us to use any Event Subscription? But PeerEventDisconnected is emitted before peers is deleted from store. So we may add new event?

I want some comments about this, @lazartravica

**Review by Kourin1996 — CHANGES_REQUESTED** [review]
Hello, Thank you for the PR. I've left some comments. It'd be great if you could check and fix if need. Thanks!

File: protocol/syncer.go (hunk 2 above)
Comment (Kourin1996) [reply protocol/syncer.go]: @munna0908 and me talked outside. I think we can pass nil to handler. But we do double check if peer is closed or not in WatchSyncWithPeer and popBlock. Can we skip p.IsClosed() in `WatchSyncWithPeer`? In this case, you modify handler little bit: if block is nill then handler return false

Comment (lazartravica) [reply protocol/syncer.go]: I agree with @Kourin1996's comments, we can skip the first `if p.IsClosed() {` in the `WatchSyncWithPeer`, and rely on the existing check in `popBlock` and add a new one to the `handler`.

@munna0908 could you also please refactor so that the callback is not called `handler` but rather `isValidator`, and avoid the double negation that we have (`return !isValidator` and `if !handler(b)`)?

File: network/server.go (hunk 3 above)
Comment (lazartravica) [reply network/server.go:193]: @munna0908 the `checkPeerConnections` can block on receiving the `PeerEventDisconnected` event and only after that check if the `s.numPeers() < MinimumPeerConnections` and try to reconnect to the bootnode.

Feel free to create a new subscription. 

Also, if you have any concerns about this (you mentioned offline somethng among the lines of cascading events) => could you care to explain how that might happen and what would be the result?

TLDR: Polling every 30 seconds is something that we should try to avoid since we have a straightforward way of subscribng to events.

**Review by lazartravica — CHANGES_REQUESTED** [review]
Please see my comments.

File: protocol/syncer.go (hunk 2 above)
Comment (munna0908) [reply protocol/syncer.go]: @lazartravica current implementation of pop block, blocks until we receive a block, but it's not safe to indefinitely wait for a block from remote node, because the remote node may be malicious. 
So in this [PR](https://github.com/0xPolygon/polygon-sdk/pull/181/commits/fbc80d4703080fd00a189c2d30c9ccaf0a5f8701) @Kourin1996  added a timeout, and he is returning a nil value, so i felt having an explict check i.e p.isClose() is required. 
Looking forward to your inputs.

Comment (munna0908) [reply protocol/syncer.go]: @lazartravica sure will refactor handler callback func

File: network/server.go (hunk 3 above)
Comment (munna0908) [reply network/server.go:193]: Fixed it Commit-[Link](https://github.com/0xPolygon/polygon-sdk/pull/186/commits/efd124d6f9bff3282e598c556dbcb01e47429b62)

File: network/server_test.go (hunk 1 above)
Comment (munna0908) [reply network/server_test.go]: @Kourin1996 Fixed it [link](https://github.com/0xPolygon/polygon-sdk/pull/186/commits/7d56b77aff6c6622aa530a9a7c47cbf686582e65)

File: network/server.go [hunk 4]
```diff
@@ -182,6 +185,32 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
...
+	}
+	for {
+
+		event, ok := <-updateCh
+		if !ok {
+			return
+		} else if event.Type == PeerEventDisconnected {
+			if s.numPeers() < MinimumPeerConnections {
+				if s.config.NoDiscover || len(s.discovery.bootnodes) == 0 {
+					//TODO: dial peers from the peerstore
```
Comment (zivkovicmilos) [inline network/server.go]: What's up with this TODO?

File: network/server.go [hunk 5]
```diff
@@ -182,6 +185,32 @@ func NewServer(logger hclog.Logger, config *Config) (*Server, error) {
...
+				} else {
+					s.dialQueue.add(s.discovery.bootnodes[rand.Intn(len(s.discovery.bootnodes))], 10)
```
Comment (zivkovicmilos) [inline network/server.go]: Can we please separate this out into multiple lines, it's messy to read? 🙈

File: network/server_test.go [hunk 6]
```diff
@@ -212,6 +214,58 @@ func TestNat(t *testing.T) {
...
+		}
+		return true, false
+
+	})
+	if err != nil {
+		return false, err
+	}
+	return res.(bool), nil
+}
+func TestPeerReconnection(t *testing.T) {
```
Comment (zivkovicmilos) [inline network/server_test.go:336]: Please provide some kind of description / test scenario for this test 🙏

**Review by zivkovicmilos — CHANGES_REQUESTED** [review]
Looks good 💯 

Left minor comments that should be resolved

File: network/server.go (hunk 4 above)
Comment (munna0908) [reply network/server.go]: @lazartravica  suggested to create a new issue in youtrack which can be handled later.

**Review by zivkovicmilos — APPROVED** [review]
Thank you for the fixes 🙏 

Looks good 🚀

**Comment by lazartravica** [comment]
We have discussed offline to bring back the polling approach without the events, to avoid code complexity.

File: protocol/syncer.go (hunk 2 above)
Comment (lazartravica) [reply protocol/syncer.go]: Thanks Rahul, I agree with you. If @Kourin1996 is on the same line, he can resolve the conversation.

**Comment by munna0908** [comment]
> We have discussed offline to bring back the polling approach without the events, to avoid code complexity.

@lazartravica reverted back to polling approach here is the [link](https://github.com/0xPolygon/polygon-sdk/pull/186/commits/50554df7736c9ff8356abd3e3c3f897bd890dce2)

File: e2e/syncer_test.go [hunk 7]
```diff
@@ -19,7 +19,7 @@ func TestSyncer(t *testing.T) {
 	// Start IBFT cluster (4 Validator + 2 Non-Validator)
 	ibftManager := framework.NewIBFTServersManager(t, IBFTMinNodes+numNonValidators, IBFTDirPrefix, func(i int, config *framework.TestServerConfig) {
 		config.SetSeal(i < IBFTMinNodes)
-		config.SetShowsLog(i == IBFTMinNodes+numNonValidators-1)
+		//config.SetShowsLog(i == IBFTMinNodes+numNonValidators-1)
```
Comment (lazartravica) [inline e2e/syncer_test.go]: Please remove this comment, we already have a PR which removed it.

**Review by lazartravica — APPROVED** [review]
Just take a look at my minor nitpick and that's it, feel free to merge

**Review by Kourin1996 — APPROVED** [review]
LGTM 👍

**Comment by munna0908** [comment]
@lazartravica 

> Just take a look at my minor nitpick and that's it, feel free to merge

Done
